torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -3781,6 +3781,56 @@ def forward(self, x):
                 else:
                     ddp(inp).sum().backward()
 
+        @require_backend({"gloo", "nccl"})
+        @require_backends_available({"gloo", "nccl"})
+        @skip_if_lt_x_gpu(2)
+        @skip_if_rocm
+        def test_ddp_device(self):
+            m = nn.Linear(10, 10).to(self.rank)
+            expected_len = 2
+
+            class TensorWrapper:
+                __slots__ = ['t']
+
+                def __init__(self, t):
+                    self.t = t
+
+            class ToyModel(torch.nn.Module):
+                def __init__(_self):  # noqa: B902
+                    super().__init__()
+                    _self.lin = nn.Linear(10, 10, bias=False)
+
+                def forward(_self, x, expected_type):  # noqa: B902
+                    # Similar to scatter, the recursive to in the single-device
+                    # case does not move tensors if they are in a custom type.
+                    self.assertTrue(isinstance(x, expected_type))
+                    if expected_type == TensorWrapper:
+                        self.assertEqual(str(x.t.device), "cpu")
+                        x.t = x.t.to(self.rank)
+                        return _self.lin(x.t)
+                    else:
+                        self.assertTrue(len(x), expected_len)
+                        self.assertTrue(x[0].device == x[1].device)
```

Comment:
Shouldn't we pass in the expected device to the constructor of ToyModel and validate it is correct here?

Reply:
We could do this, but the test is basically expected to validate that the device is the current rank of the process, as we are testing single GPU per process. The next line asserts that the input is on the expected device.